File: apps/kira-studio/internal/httpclient/body.go
// body.go is P3's own file: one serializer per request-body mode (D2/D5), turning a Body into
// what net/http needs — a reader, a working GetBody for 307/308 replay (F4), an exact
// Content-Length so nothing here is ever sent chunked (F5), and the mode's default Content-Type
// (client.go's Send applies D7's precedence over it). No dependency (D1): mime/multipart's own
// deterministic framing for a fixed boundary (F15) is what makes the multipart dry run possible.
package httpclient

import (
	"fmt"
	"io"
	"mime/multipart"
	"net/textproto"
	"net/url"
	"os"
	"path/filepath"
	"strings"
)

// BodyMode is this app's own `mode` spelling, originally Postman's own (D2) before the raw/code
// split — "file" is what the UI calls binary.
type BodyMode string

const (
	BodyNone       BodyMode = "none"
	BodyRaw        BodyMode = "raw"
	BodyCode       BodyMode = "code"
	BodyURLEncoded BodyMode = "urlencoded"
	BodyFormData   BodyMode = "formdata"
	// BodyFile — one local file as the entire body; the UI calls this **binary**.
	BodyFile BodyMode = "file"
)

// validBodyModes — D12's parity guard reads this exact map[string]bool literal as plain text
// (extractGoStringSet), so it stays a literal rather than being derived from the consts above.
var validBodyModes = map[string]bool{
	"none": true, "raw": true, "code": true, "urlencoded": true, "formdata": true, "file": true,
}

// contentTypeByCodeLanguage — D7's default Content-Type per `code` mode sub-language. A
// map[string]string literal, not a switch, specifically so D12's parity test
// (extractGoStringMap) can read it as plain text against CONTENT_TYPE_BY_CODE_LANGUAGE (http.ts)
// — do not "simplify" this into a switch. Plain `raw` always sends text/plain and needs no table.
var contentTypeByCodeLanguage = map[string]string{
	"javascript": "application/javascript",
	"json":       "application/json",
	"html":       "text/html",
	"xml":        "application/xml",
}

// Field is one urlencoded row (D5).
type Field struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

// FormField is one form-data row (D5). Kind == "file" means Path is an absolute local path (D4)
// and Value is ignored; Kind == "text" means Value is the field's content and Path is ignored.
// ContentType is a per-part override; "" means the per-mode default (the row's own Content type
// field when set, else "application/octet-stream" for a file part, nothing for a text part).
type FormField struct {
	Name        string `json:"name"`
	Kind        string `json:"kind"`
	Value       string `json:"value"`
	Path        string `json:"path"`
	ContentType string `json:"contentType"`
}

// Body is a tagged union: Mode selects which member is meaningful and every other member is
// ignored (D5). Raw is the plain-text buffer (raw mode only); Code/CodeLanguage are the
// syntax-highlighted buffer and its language (code mode only) — two separate buffers so switching
// between the two modes never loses either one's text.
type Body struct {
	Mode         string      `json:"mode"`
	Raw          string      `json:"raw"`
	Code         string      `json:"code"`
	CodeLanguage string      `json:"codeLanguage"`
	URLEncoded   []Field     `json:"urlEncoded"`
	FormData     []FormField `json:"formData"`
	File         string      `json:"file"`
}

// buildBody turns a Body into what net/http needs. contentType is the default this mode implies
// (D7 decides whether Send actually applies it); length is exact, never -1, so nothing here is
// ever sent chunked (F5). formBoundary is the boundary a formdata body must use — client.go's
// Content-Type precedence (D7) resolves it (the user's own header's boundary parameter, or a
// freshly minted one) before calling this; every other mode ignores it.
func buildBody(b Body, formBoundary string) (
	body io.ReadCloser, getBody func() (io.ReadCloser, error), length int64, contentType string, err error,
) {
	mode := b.Mode
	if mode == "" {
		// The zero-value Body{} (every httpclient.Request built with no Body field set at all,
		// e.g. every pre-P3 test in this package) means "no body" — the same default HasBody's
		// own zero value (false) gave P2.
		mode = string(BodyNone)
	}
	if !validBodyModes[mode] {
		return nil, nil, 0, "", newError(CodeBadRequest, "unknown body mode: "+b.Mode, nil)
	}
	switch BodyMode(mode) {
	case BodyNone:
		return nil, nil, 0, "", nil
	case BodyRaw:
		return buildRaw(b.Raw)
	case BodyCode:
		return buildCode(b.Code, b.CodeLanguage)
	case BodyURLEncoded:
		return buildURLEncoded(b.URLEncoded)
	case BodyFile:
		return buildFile(b.File)
	case BodyFormData:
		return buildFormData(b.FormData, formBoundary)
	}
	return nil, nil, 0, "", newError(CodeBadRequest, "unknown body mode: "+b.Mode, nil)
}

// buildRaw is plain text only — no sub-language, always text/plain.
func buildRaw(raw string) (io.ReadCloser, func() (io.ReadCloser, error), int64, string, error) {
	reader := func() io.ReadCloser { return io.NopCloser(strings.NewReader(raw)) }
	return reader(), func() (io.ReadCloser, error) { return reader(), nil },
		int64(len(raw)), "text/plain", nil
}

// buildCode is the syntax-highlighted sibling of buildRaw — same shape, its Content-Type depends
// on which of the four code languages was chosen.
func buildCode(code, codeLanguage string) (io.ReadCloser, func() (io.ReadCloser, error), int64, string, error) {
	reader := func() io.ReadCloser { return io.NopCloser(strings.NewReader(code)) }
	return reader(), func() (io.ReadCloser, error) { return reader(), nil },
		int64(len(code)), contentTypeByCodeLanguage[codeLanguage], nil
}

// buildURLEncoded hand-writes the encoder rather than reusing url.Values.Encode (F6/D1): that
// sorts keys, silently rewriting a user's field order, and cannot express a disabled row (the
// renderer already dropped those before this ever runs, D5). url.QueryEscape encodes a space as
// '+', which is correct in an application/x-www-form-urlencoded body — the deliberate opposite of
// views/httprequest/url.ts's buildQuery, which encodeURIComponents a query-string space as '%20'.
// The two encoders differ on purpose; neither is "fixed" into the other.
// encodeURLEncodedFields is buildURLEncoded's own encoder, factored out so wire.go's P9 D4
// rendering can render the *encoded* string — the one that actually goes out — without a second
// encoder that could drift from this one.
func encodeURLEncodedFields(fields []Field) (string, error) {
	parts := make([]string, 0, len(fields))
	for _, f := range fields {
		if strings.TrimSpace(f.Name) == "" {
			return "", newError(CodeBadRequest, "a urlencoded field is missing a name", nil)
		}
		parts = append(parts, url.QueryEscape(f.Name)+"="+url.QueryEscape(f.Value))
	}
	return strings.Join(parts, "&"), nil
}

func buildURLEncoded(fields []Field) (io.ReadCloser, func() (io.ReadCloser, error), int64, string, error) {
	encoded, err := encodeURLEncodedFields(fields)
	if err != nil {
		return nil, nil, 0, "", err
	}
	reader := func() io.ReadCloser { return io.NopCloser(strings.NewReader(encoded)) }
	return reader(), func() (io.ReadCloser, error) { return reader(), nil },
		int64(len(encoded)), "application/x-www-form-urlencoded", nil
}

// buildFile is the binary body: one whole local file, streamed by Go and never by the renderer
// (D4). Postman sets no Content-Type for this mode (F3), and neither do we. Re-`os.Stat`s at send
// time rather than trusting a stored size — the file may have changed or vanished since it was
// picked (D4).
func buildFile(path string) (io.ReadCloser, func() (io.ReadCloser, error), int64, string, error) {
	trimmed := strings.TrimSpace(path)
	if trimmed == "" {
		return nil, nil, 0, "", newError(CodeBadRequest, "no file chosen for the request body", nil)
	}
	info, err := os.Stat(trimmed)
	if err != nil {
		return nil, nil, 0, "", newError(CodeBadRequest, "could not read local file "+trimmed+": "+err.Error(), err)
	}
	if info.IsDir() {
		return nil, nil, 0, "", newError(CodeBadRequest, trimmed+" is a directory, not a file", nil)
	}
	f, err := os.Open(trimmed)
	if err != nil {
		return nil, nil, 0, "", newError(CodeBadRequest, "could not read local file "+trimmed+": "+err.Error(), err)
	}
	getBody := func() (io.ReadCloser, error) {
		reopened, openErr := os.Open(trimmed)
		if openErr != nil {
			return nil, openErr
		}
		return reopened, nil
	}
	return f, getBody, info.Size(), "", nil
}

// formPart pairs a FormField with the file size resolved once, at prepare time — reused for both
// the dry-run count and the real stream (and any later GetBody replay), so a redirect's replayed
// body reports the same Content-Length the first attempt did even if the file changed size in
// between (D6).
type formPart struct {
	field FormField
	size  int64 // meaningful only when field.Kind == "file"
}

func prepareFormParts(fields []FormField) ([]formPart, error) {
	parts := make([]formPart, 0, len(fields))
	for _, f := range fields {
		if strings.TrimSpace(f.Name) == "" {
			return nil, newError(CodeBadRequest, "a form-data field is missing a name", nil)
		}
		if f.Kind != "file" {
			parts = append(parts, formPart{field: f})
			continue
		}
		trimmed := strings.TrimSpace(f.Path)
		if trimmed == "" {
			return nil, newError(CodeBadRequest, "no file chosen for form-data field "+f.Name, nil)
		}
		info, err := os.Stat(trimmed)
		if err != nil {
			return nil, newError(CodeBadRequest, "could not read local file "+trimmed+": "+err.Error(), err)
		}
		if info.IsDir() {
			return nil, newError(CodeBadRequest, trimmed+" is a directory, not a file", nil)
		}
		parts = append(parts, formPart{field: f, size: info.Size()})
	}
	return parts, nil
}

// quoteEscaper mirrors mime/multipart's own unexported one (used by CreateFormFile) — the package
// gives no other way to build a Content-Disposition header for a caller-supplied Content-Type.
var quoteEscaper = strings.NewReplacer("\\", "\\\\", `"`, "\\\"")

func formPartHeader(f FormField) textproto.MIMEHeader {
	h := make(textproto.MIMEHeader)
	if f.Kind == "file" {
		filename := filepath.Base(f.Path)
		h.Set("Content-Disposition", fmt.Sprintf(`form-data; name="%s"; filename="%s"`,
			quoteEscaper.Replace(f.Name), quoteEscaper.Replace(filename)))
		ct := f.ContentType
		if ct == "" {
			ct = "application/octet-stream"
		}
		h.Set("Content-Type", ct)
	} else {
		h.Set("Content-Disposition", fmt.Sprintf(`form-data; name="%s"`, quoteEscaper.Replace(f.Name)))
		if f.ContentType != "" {
			h.Set("Content-Type", f.ContentType)
		}
	}
	return h
}

// countWriter measures exactly what a real multipart.Writer pass would produce, without holding
// any of it in memory (F15).
type countWriter struct{ n int64 }

func (c *countWriter) Write(p []byte) (int, error) {
	c.n += int64(len(p))
	return len(p), nil
}

// multipartLength is the dry run (F15/D6): the identical boundary and part headers as the real
// pass, written to a counting writer; a file part adds its already-`os.Stat`'d size instead of
// being read. Same boundary, same headers, same terminator ⇒ the count the real pass will produce.
func multipartLength(parts []formPart, boundary string) (int64, error) {
	cw := &countWriter{}
	mw := multipart.NewWriter(cw)
	if err := mw.SetBoundary(boundary); err != nil {
		return 0, newError(CodeBadRequest, "invalid multipart boundary: "+err.Error(), err)
	}
	for _, p := range parts {
		part, err := mw.CreatePart(formPartHeader(p.field))
		if err != nil {
			return 0, newError(CodeHTTPTransport, "could not build multipart body: "+err.Error(), err)
		}
		if p.field.Kind == "file" {
			cw.n += p.size
			continue
		}
		if _, err := part.Write([]byte(p.field.Value)); err != nil {
			return 0, newError(CodeHTTPTransport, "could not build multipart body: "+err.Error(), err)
		}
	}
	if err := mw.Close(); err != nil {
		return 0, newError(CodeHTTPTransport, "could not build multipart body: "+err.Error(), err)
	}
	return cw.n, nil
}

// streamFormData is the real pass: written into an io.Pipe by one goroutine so the caller can
// start reading before every part is ready, streaming each file straight from disk (D4) rather
// than buffering it. io.CopyN with the size prepareFormParts already resolved means a file that
// shrank in between fails loudly (io.ErrUnexpectedEOF, propagated via CloseWithError) and a file
// that grew is truncated to the counted size — either way the bytes sent match Content-Length
// (D6). The goroutine never outlives the request: it exits on its own first write error, and
// cancelling the request closes pr, which turns the next write into io.ErrClosedPipe.
func streamFormData(parts []formPart, boundary string) *io.PipeReader {
	pr, pw := io.Pipe()
	go func() {
		var werr error
		defer func() { _ = pw.CloseWithError(werr) }()

		mw := multipart.NewWriter(pw)
		if err := mw.SetBoundary(boundary); err != nil {
			werr = err
			return
		}
		for _, p := range parts {
			part, err := mw.CreatePart(formPartHeader(p.field))
			if err != nil {
				werr = err
				return
			}
			if p.field.Kind == "file" {
				f, openErr := os.Open(p.field.Path)
				if openErr != nil {
					werr = openErr
					return
				}
				_, copyErr := io.CopyN(part, f, p.size)
				closeErr := f.Close()
				if copyErr != nil {
					werr = copyErr
					return
				}
				if closeErr != nil {
					werr = closeErr
					return
				}
				continue
			}
			if _, err := part.Write([]byte(p.field.Value)); err != nil {
				werr = err
				return
			}
		}
		werr = mw.Close()
	}()
	return pr
}

// mintBoundary hands back a fresh random boundary with no part written — multipart.NewWriter
// generates one immediately, before anything is written to the (discarded) underlying writer.
func mintBoundary() string {
	return multipart.NewWriter(io.Discard).Boundary()
}

func buildFormData(fields []FormField, boundary string) (
	io.ReadCloser, func() (io.ReadCloser, error), int64, string, error,
) {
	parts, err := prepareFormParts(fields)
	if err != nil {
		return nil, nil, 0, "", err
	}
	if boundary == "" {
		boundary = mintBoundary()
	}
	length, err := multipartLength(parts, boundary)
	if err != nil {
		return nil, nil, 0, "", err
	}
	contentType := "multipart/form-data; boundary=" + boundary
	getBody := func() (io.ReadCloser, error) { return streamFormData(parts, boundary), nil }
	return streamFormData(parts, boundary), getBody, length, contentType, nil
}

File: packages/api-core/src/http/body.ts
import type { EditorLanguageId } from '@kira/shared/domain/editor';
import {
  CONTENT_TYPE_BY_CODE_LANGUAGE,
  type HttpBodyMode,
  type HttpCodeLanguage,
  type HttpHeaderState,
  type HttpRequestTabState,
} from '@kira/shared/domain/http';

// C5/D9: the five body-mode labels, in the builder's own order. `title` is the tooltip F12 leans
// on instead of widening SegmentedControl for this one caller's label widths.
export const BODY_MODE_OPTIONS: readonly {
  value: HttpBodyMode;
  label: string;
  title: string;
  testid: string;
}[] = [
  { value: 'none', label: 'none', title: 'No request body', testid: 'http-body-mode-none' },
  {
    value: 'formdata',
    label: 'form-data',
    title: 'Multipart fields, including real files',
    testid: 'http-body-mode-formdata',
  },
  {
    value: 'urlencoded',
    label: 'x-www-form-urlencoded',
    title: 'URL-encoded form fields',
    testid: 'http-body-mode-urlencoded',
  },
  {
    value: 'raw',
    label: 'raw',
    title: 'Plain text',
    testid: 'http-body-mode-raw',
  },
  {
    value: 'code',
    label: 'Code',
    title: 'JavaScript, JSON, HTML or XML',
    testid: 'http-body-mode-code',
  },
  {
    value: 'file',
    label: 'binary',
    title: "One local file, sent as the request's entire body",
    testid: 'http-body-mode-file',
  },
];

// The `code` mode's sub-selector, in the same order raw's own sub-selector used to list them
// (minus Text, which is what plain `raw` now means).
export const CODE_LANGUAGE_OPTIONS: readonly { value: HttpCodeLanguage; label: string }[] = [
  { value: 'javascript', label: 'JavaScript' },
  { value: 'json', label: 'JSON' },
  { value: 'html', label: 'HTML' },
  { value: 'xml', label: 'XML' },
];

// Which CodeMirror grammar a code sub-language renders with — only json/xml have a real grammar
// in this app (languages.ts); javascript/html render unhighlighted, exactly as they did as raw
// sub-languages before the split (no requirement beyond colour for either).
const CODE_LANGUAGE_TO_EDITOR: Readonly<Record<HttpCodeLanguage, EditorLanguageId>> = {
  javascript: 'plain',
  json: 'json',
  html: 'plain',
  xml: 'xml',
};

export function editorLanguageForCode(codeLanguage: HttpCodeLanguage): EditorLanguageId {
  return CODE_LANGUAGE_TO_EDITOR[codeLanguage];
}

/** The user's own Content-Type header value, if any enabled row sets one — case-insensitive,
 *  last one wins (mirrors client.go's own headerValue). */
export function userContentTypeHeader(headers: readonly HttpHeaderState[]): string | undefined {
  let value: string | undefined;
  for (const h of headers) {
    if (h.enabled && h.name.trim().toLowerCase() === 'content-type') value = h.value;
  }
  return value;
}

/** D7: the Content-Type Go's own default would apply for this mode — "" when the mode sends none
 *  at all (none, binary). Mirrors internal/httpclient/client.go's own Send precedence, read-only:
 *  this never talks to Go, it only states what D7 documents Go will do. */
export function defaultContentTypeFor(mode: HttpBodyMode, codeLanguage: HttpCodeLanguage): string {
  switch (mode) {
    case 'none':
    case 'file':
      return '';
    case 'raw':
      return 'text/plain';
    case 'code':
      return CONTENT_TYPE_BY_CODE_LANGUAGE[codeLanguage];
    case 'urlencoded':
      return 'application/x-www-form-urlencoded';
    case 'formdata':
      return 'multipart/form-data; boundary=…';
  }
}

/** D9's caption: exactly what Content-Type the send will actually carry — a hand-set header wins
 *  (D7's precedence), binary gets the explicit "no Content-Type" sentence (F3: Postman sets none
 *  either), and 'none' has nothing to caption. */
export function contentTypeCaption(
  mode: HttpBodyMode,
  codeLanguage: HttpCodeLanguage,
  userContentType: string | undefined,
): string {
  const trimmed = (userContentType ?? '').trim();
  if (trimmed !== '') return `Content-Type: ${trimmed} (from your header)`;
  if (mode === 'none') return '';
  if (mode === 'file') return 'No Content-Type (binary)';
  return `Content-Type: ${defaultContentTypeFor(mode, codeLanguage)} (auto)`;
}

/** D9: the Body segment's own count badge — mirrors REQUEST_PANE_OPTIONS's existing
 *  "bake the count into the label" technique for Params/Headers (HttpRequestView.vue). */
export function bodyBadgeLabel(state: HttpRequestTabState): string {
  switch (state.bodyMode) {
    case 'none':
      return 'Body';
    case 'raw':
      return 'Body (raw)';
    case 'code':
      return 'Body (code)';
    case 'urlencoded': {
      const n = state.urlEncoded.filter((f) => f.enabled && f.name.trim() !== '').length;
      return n > 0 ? `Body (${n})` : 'Body';
    }
    case 'formdata': {
      const n = state.formData.filter((f) => f.enabled && f.name.trim() !== '').length;
      return n > 0 ? `Body (${n})` : 'Body';
    }
    case 'file':
      return state.binaryFile ? 'Body (1 file)' : 'Body';
  }
}
